server/controller/src/main/java/ai/starwhale/mlops/domain/runtime/RuntimeService.java

```diff
@@ -462,15 +491,150 @@ public void pull(String projectUrl, String runtimeUrl, String versionUrl, HttpSe
+                throw new SwProcessException(ErrorType.INFRA,
+                        "deploying job for image build error:" + k8sE.getMessage());
+            }
+        }
+    }
+
+    private boolean isExist(String runtimeUrl, String versionUrl, RuntimeVersionEntity runtimeVersion) {
+        try {
+            var existJob = k8sClient.getJob(runtimeVersion.getVersionName());
+            if (null != existJob) {
+                return true;
+            }
+        } catch (ApiException k8sE) {
+            if (k8sE.getCode() != HttpServletResponse.SC_NOT_FOUND) {
+                log.error("k8s api invoke error {}", k8sE.getResponseBody(), k8sE);
+                throw new SwProcessException(ErrorType.INFRA, "k8s api invoke error" + k8sE.getMessage());
+            }
+            // go on
+        }
+        return false;
+    }
+
+    private void checkDockerRegistry() {
+        try {
+            // judge registry secret whether has been created
+            k8sClient.getSecret(DOCKER_REGISTRY_SECRET);
+        } catch (ApiException k8sE) {
+            if (k8sE.getCode() == HttpServletResponse.SC_NOT_FOUND) {
+                log.debug("secret is not found {}, now to build it", k8sE.getResponseBody(), k8sE);
+                var sysSetting = systemSettingService.getSystemSetting();
+                if (null != sysSetting
+                        && null != sysSetting.getDockerSetting()
+                        && null != sysSetting.getDockerSetting().getRegistry()
+                        && null != sysSetting.getDockerSetting().getUserName()
+                        && null != sysSetting.getDockerSetting().getPassword()) {
+                    try {
+                        var dockerConfig = Map.of("auths", Map.of(
+                                sysSetting.getDockerSetting().getRegistry(), Map.of(
+                                "username", sysSetting.getDockerSetting().getUserName(),
+                                "password", sysSetting.getDockerSetting().getPassword())
+                        ));
+                        k8sClient.createSecret(new V1Secret()
+                                .metadata(new V1ObjectMeta().name(DOCKER_REGISTRY_SECRET))
+                                .type("kubernetes.io/dockerconfigjson")
+                                .stringData(Map.of(".dockerconfigjson", JSONUtil.toJsonStr(dockerConfig)))
+                        );
+                    } catch (Exception e) {
+                        log.error("create secret error {}", e.getMessage(), e);
+                        throw new SwValidationException(ValidSubject.RUNTIME,
+                            "create secret error when building image for runtime");
+                    }
+
+                } else {
+                    throw new SwValidationException(ValidSubject.RUNTIME,
+                        "can't found docker registry info, please set it in system setting.");
+                }
+            }
+            // go on
+        }
```

Comment:
It may not be possible to handle the scenario where the user updates authentication information for data that already exists. 
Can we write to k8s every time the app starts or the user updates the configuration?

Reply:
done, update secret when user update settings every times